scalanet/src/io/iohk/scalanet/peergroup/dynamictls/DynamicTLSPeerGroup.scala

```diff
@@ -48,8 +48,6 @@ class DynamicTLSPeerGroup[M] private (val config: Config)(
 ) extends TerminalPeerGroup[PeerInfo, M]
     with StrictLogging {
 
-  protected override val s = scheduler
-
   private val sslServerCtx: SslContext = DynamicTLSPeerGroupUtils.buildCustomSSlContext(SSLContextForServer, config)
 
   private val serverQueue = CloseableQueue.unbounded[ServerEvent[PeerInfo, M]](ChannelType.SPMC).runSyncUnsafe()
```

Comment:
I am wondering, that if maybe with currrent setup we should have this queue as `ChannelType.MPMC` ? From what i can see all events to it are pushed from `implicit scheduler: Scheduler` provided by user to class, and there is no guarannte that user of the class will provide one threaded scheduler.

Other solution (more neat in my opinion) would be to create one threaded `NioEventLoopGroup` for accepting connections and pushing sever events on `serverQueue`, then this external scheduler could be deleted completely and all producing would happen on netty threads. Then you would have :
```
  private val workerGroup = new NioEventLoopGroup()
  private val acceptinGroup = new NioEventLoopGroup(1)
  
  private val serverBootstrap = new ServerBootstrap()
    .group(acceptinGroup, workerGroup)
```

wdyt ?

Reply:
My current understanding is that `SPMC` is fine because of what I pointed to in this comment: https://github.com/input-output-hk/scalanet/pull/92#discussion_r501611101

Basically I think that unless you actually publish from multiple threads at the same time, you are fine with the "single producer" scenario, which in our case is true because:
* we [push server events](https://github.com/input-output-hk/scalanet/blob/717cc24572a58f7548a7af0a4745791ac2373100/scalanet/src/io/iohk/scalanet/peergroup/dynamictls/DynamicTLSPeerGroupInternals.scala#L267) with `.runSyncUnsafe()` which is running on the external scheduler but the thread accepting netty connections one will wait for it, so it won't try to offer anything else at the same time
* we [push messages](https://github.com/input-output-hk/scalanet/blob/717cc24572a58f7548a7af0a4745791ac2373100/scalanet/src/io/iohk/scalanet/peergroup/dynamictls/DynamicTLSPeerGroupInternals.scala#L45-L46) on a single threaded scheduler on top of the netty one as you suggested, so again no chance to do two at the same time

I actually started to remove these schedulers, but then gave up when I realised they are present mainly so that we can block creating the queue instance. Ideally IMO we could refactor the TLS group to look more like the `StaticUDPPeerGroup` and do almost everything in tasks, async, on top of the netty channel.

Reply:
this the case i am worried about:

> we push server events with .runSyncUnsafe() which is running on the external scheduler but the thread accepting netty connections one will wait for it, so it won't try to offer anything else at the same time

as each connection currently is accepted on possibly different thread. So there can be potentially mulitple handshakes in progress, an if lets say two of them end up in the same time then both can run `.runSyncUnsafe()` and if external scheduler is muliple threaded then we will have multiple producers.

Reply:
Hm, I thought Netty only has 1 boss thread where it accepts connections. I see that we use the same work group for accepting connections and handling messages, but but still I thought it will always the be same thread doing a particular task.

Reply:
https://stackoverflow.com/questions/5474372/how-netty-uses-thread-pools says there's only one boss thread per socket channel.

Reply:
Hm yep this thread may be right, but when we talk about accepting connection we are talking about tcp handshake. When we are push our success event:

Reply:
That's a good point, I didn't realise that, thanks! 

Changed it to MPMC.